Kernel/FileSystem/FATFileSystem.cpp

```diff
@@ -0,0 +1,361 @@
+
+        BlockBasedFileSystem::BlockIndex first_block = fs().first_block_of_cluster(cluster);
+        for (u8 i = 0; i < fs().boot_record()->sectors_per_cluster; i++)
+            block_list.append(BlockBasedFileSystem::BlockIndex { first_block.value() + i });
+
+        u32 fat_offset = cluster * 4;
+        u32 fat_sector_index = fs().boot_record()->reserved_sector_count + (fat_offset / fs().m_logical_block_size);
+        u32 entry_offset = fat_offset % fs().m_logical_block_size;
+
+        TRY(fs().raw_read(fat_sector_index, fat_sector_buffer));
+
+        cluster = *(u32*)&fat_sector->data()[entry_offset];
+        cluster &= 0x0FFFFFFF;
+    }
+
+    return block_list;
+}
+
+ErrorOr<NonnullOwnPtr<KBuffer>> FATInode::read_block_list()
+{
+    VERIFY(m_inode_lock.is_locked());
+
+    dbgln_if(FAT_DEBUG, "FATFS: reading block list for inode {} ({} blocks)", index(), m_block_list.size());
+
+    if (m_block_list.is_empty())
+        m_block_list = TRY(compute_block_list());
+
+    auto builder = TRY(KBufferBuilder::try_create());
+
+    u8 buffer[512];
+    VERIFY(fs().m_logical_block_size <= sizeof(buffer));
+    auto buf = UserOrKernelBuffer::for_kernel_buffer(buffer);
+
+    for (BlockBasedFileSystem::BlockIndex block : m_block_list) {
+        dbgln_if(FAT_VERY_DEBUG, "FATFS: reading block: {}", block);
+        TRY(fs().raw_read(block, buf));
+        TRY(builder.append((char const*)buffer, fs().m_logical_block_size));
+    }
+
+    auto blocks = builder.build();
+    if (!blocks)
+        return ENOMEM;
+    return blocks.release_nonnull();
+}
+
+ErrorOr<Vector<RefPtr<FATInode>>> FATInode::traverse()
+{
+    VERIFY(m_entry.attributes & FATAttributes::Directory);
+
+    Vector<RefPtr<FATInode>> inodes;
+    Vector<FATLongFileNameEntry> lfn_entries;
+    auto blocks = TRY(read_block_list());
+
+    for (u32 i = 0; i < blocks->size() / sizeof(FATEntry); i++) {
+        auto* entry = reinterpret_cast<FATEntry*>(blocks->data() + i * sizeof(FATEntry));
+        if (entry->filename[0] == 0x00) {
+            dbgln_if(FAT_VERY_DEBUG, "FATFS: Found end entry");
+            return inodes;
+        } else if (entry->filename[0] == 0xE5) {
+            dbgln_if(FAT_VERY_DEBUG, "FATFS: Found unused entry");
+        } else if (entry->attributes == 0x0F) {
+            dbgln_if(FAT_VERY_DEBUG, "FATFS: Found LFN entry");
+            lfn_entries.append(*(FATLongFileNameEntry*)entry);
+        } else {
+            dbgln_if(FAT_VERY_DEBUG, "FATFS: Found 8.3 entry");
+            lfn_entries.reverse();
+            auto inode = FATInode::create(fs(), *entry, lfn_entries);
+            inodes.append(inode);
+            lfn_entries.clear();
+        }
+    }
+
+    return EINVAL;
+}
+
+void FATInode::compute_filename(Vector<FATLongFileNameEntry> lfn_entries)
+{
+    VERIFY(!m_filename);
+
+    // FIXME: Find a way to handle errors here instead of using MUST
```

Comment:
This function could probably already be `ErrorOr<void>`, or at least I don't see a reason why it couldn't be. Of course, this only moves the "handling errors" part further down the line, but it's a first step.

Reply:
The only place that function is called is the constructor which is not failable. I made the function `ErrorOr<void>` and moved that `FIXME` to the constructor, but I don't know if that's a good solution.